### PR TITLE
Prescription table

### PR DESCRIPTION
**Pull request by Bm2pires**
This Pull Request includes the issue prescription functionality for staff (Doctors and Nurses) as well as the prescription table.

### CODE REVIEW

**Comment by Tom-Goring** [comment]
Rebase onto master successfully

File: src/main/java/Controllers/IssuePrescriptionServlet.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * Processes requests for both HTTP <code>GET</code> and <code>POST</code>
+     * methods.
+     *
+     * @param request servlet request
+     * @param response servlet response
+     * @throws ServletException if a servlet-specific error occurs
+     * @throws IOException if an I/O error occurs
+     */
+    protected void processRequest(HttpServletRequest request, HttpServletResponse response)
```
Comment (Tom-Goring) [inline src/main/java/Controllers/IssuePrescriptionServlet.java:36]: This method should be removed, it is only called once in the doGet method.

File: src/main/java/Controllers/IssuePrescriptionServlet.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+     *
+     * @param request servlet request
+     * @param response servlet response
+     * @throws ServletException if a servlet-specific error occurs
+     * @throws IOException if an I/O error occurs
+     */
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException {
+        processRequest(request, response);
```
Comment (Tom-Goring) [inline src/main/java/Controllers/IssuePrescriptionServlet.java:65]: See the above comment

File: src/main/java/Controllers/IssuePrescriptionServlet.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+     * @param request servlet request
+     * @param response servlet response
+     * @throws ServletException if a servlet-specific error occurs
+     * @throws IOException if an I/O error occurs
+     */
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        //catch all user input and pass into DAO
+        
+        Prescription prescription = new Prescription(Integer.parseInt(request.getParameter("submitted-patientid")),
```
Comment (Tom-Goring) [inline src/main/java/Controllers/IssuePrescriptionServlet.java:80]: We probably don't want Users to be forced to somehow find out UserIDs in order to issue prescriptions. A drop down of patients would be better. See the appointments and approvals code to see how to do a <select> of names where the value of each name is the userID.

File: src/main/java/Controllers/LoginServlet.java [hunk 4]
```diff
@@ -19,7 +21,8 @@
 @WebServlet(name = "LoginServlet")
 public class LoginServlet extends HttpServlet {
     protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-
+        
+//        Tables.recreateTables();
```
Comment (Tom-Goring) [inline src/main/java/Controllers/LoginServlet.java:25]: Changes to this file should be reverted

File: src/main/java/Models/Prescription/Prescription.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @author Bredan
+ */
+
+public class Prescription {
+    private final Integer patientID;
+    private final String medicine;
+    private final Integer quantity;
+    private final Boolean repeating;
+    private final String issueDate;
```
Comment (Tom-Goring) [inline src/main/java/Models/Prescription/Prescription.java:13]: This should be an actual Date type, not a string

File: src/main/java/Models/Prescription/Prescription.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    private final String endDate;
```
Comment (Tom-Goring) [inline src/main/java/Models/Prescription/Prescription.java:14]: Same for this type

File: src/main/java/Models/Prescription/Prescription.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
+package Models.Prescription;
+
+/**
+ *
+ * @author Bredan
+ */
+
+public class Prescription {
+    private final Integer patientID;
```
Comment (Tom-Goring) [inline src/main/java/Models/Prescription/Prescription.java:9]: This can be a User type called `patient` instead of a patientID

File: src/main/java/Models/Prescription/PrescriptionDAO.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        Connection con = Database.getInstance().getConnection();
+        
+        try {
+            PreparedStatement ps = con.prepareStatement("INSERT INTO PRESCRIPTION (PATIENTID, MEDICINE, QUANTITY, REPEATING, DATE_ISSUED, END_DATE) VALUES (?, ?, ?, ?, ?, ?)");
+            ps.setInt(1, prescription.getPatientID());
+            ps.setString(2, prescription.getMedicine());
+            ps.setInt(3, prescription.getQuantity());
+            ps.setBoolean(4, prescription.getRepeating());
+            
+            java.sql.Date issueDate = java.sql.Date.valueOf(prescription.getIssueDate());
```
Comment (Tom-Goring) [inline src/main/java/Models/Prescription/PrescriptionDAO.java:28]: See the above comments about the date types in the Prescription class.

File: src/main/java/Utils/Tables.java [hunk 9]
```diff
@@ -39,6 +39,20 @@ public static void createUserTable() {
             );
             UserDao.insertUser(admin);
             System.out.println("Admin user created successfully!");
+            
+            // staff test user
+            Tuple<byte[], byte[]> saltAndHash1 = createSaltAndHash("staff");
```
Comment (Tom-Goring) [inline src/main/java/Utils/Tables.java:44]: This will need to be removed - we only want to create a base admin class, none others. Staff will be added after deployment.

File: src/main/webapp/IssuePrescription.jsp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+<html>
+    <head>
+        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+        <title>Issue Prescription Form</title>
+    </head>
+    <body>
+        <h1>Issue Prescription</h1>
+        <form method="post" action="${pageContext.request.contextPath}/IssuePrescription">
+            <label>
+                Patient ID: <input name="submitted-patientid">
```
Comment (Tom-Goring) [inline src/main/webapp/IssuePrescription.jsp:18]: As per my other comment, look at the appointment and approvals pages to see how to do a combo-box with the User names in it for the user, which submits User IDs when submitted.

Comment (Tom-Goring) [reply src/main/webapp/IssuePrescription.jsp:18]: You should also probably add a `required` to these.

File: src/main/webapp/WEB-INF/web.xml [hunk 11]
```diff
@@ -1,10 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-
-<web-app xmlns="http://java.sun.com/xml/ns/javaee"
-	 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-	 xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd"
-	 version="3.0">
-    
+<web-app version="3.0" xmlns="http://java.sun.com/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd">
```
Comment (Tom-Goring) [inline src/main/webapp/WEB-INF/web.xml:2]: Should probably revert the change to these lines

File: src/main/webapp/WEB-INF/web.xml [hunk 12]
```diff
@@ -27,11 +22,11 @@
         <servlet-name>LoginServlet</servlet-name>
         <servlet-class>Controllers.LoginServlet</servlet-class>
     </servlet>
-   <servlet-mapping>
-       <servlet-name>LoginServlet</servlet-name>
-       <url-pattern>/login</url-pattern>
-   </servlet-mapping>
-
+    <servlet-mapping>
```
Comment (Tom-Goring) [inline src/main/webapp/WEB-INF/web.xml:25]: Need to remove these whitespace changes

**Review by Tom-Goring — CHANGES_REQUESTED** [review]
Left some comments about changes that need to be made - generally looks good, good work!

File: src/main/java/Controllers/IssuePrescriptionServlet.java (hunk 1 above)
Comment (Bm2pires) [reply src/main/java/Controllers/IssuePrescriptionServlet.java:36]: Just deleted this function!

File: src/main/java/Controllers/IssuePrescriptionServlet.java (hunk 2 above)
Comment (Bm2pires) [reply src/main/java/Controllers/IssuePrescriptionServlet.java:65]: Deleted this also!

File: src/main/java/Controllers/LoginServlet.java (hunk 4 above)
Comment (Bm2pires) [reply src/main/java/Controllers/LoginServlet.java:25]: Reverted these changes!

File: src/main/java/Models/Prescription/Prescription.java (hunk 5 above)
Comment (Bm2pires) [reply src/main/java/Models/Prescription/Prescription.java:13]: image
it is converted to date type in the PrescriptionDAO

File: src/main/java/Models/Prescription/Prescription.java (hunk 6 above)
Comment (Bm2pires) [reply src/main/java/Models/Prescription/Prescription.java:14]: this type is also converted in the prescriptionDAO. Although not explicitly converted to date in the prescription object.

File: src/main/webapp/WEB-INF/web.xml (hunk 12 above)
Comment (Bm2pires) [reply src/main/webapp/WEB-INF/web.xml:25]: should be fixed now!

File: src/main/webapp/WEB-INF/web.xml (hunk 11 above)
Comment (Bm2pires) [reply src/main/webapp/WEB-INF/web.xml:2]: Reverted!

File: src/main/webapp/IssuePrescription.jsp (hunk 10 above)
Comment (Bm2pires) [reply src/main/webapp/IssuePrescription.jsp:18]: required tags added in!

File: src/main/java/Utils/Tables.java (hunk 9 above)
Comment (Bm2pires) [reply src/main/java/Utils/Tables.java:44]: I've only kept this in for testing purposes. Will be deleted however when possible

File: src/main/java/Models/Prescription/Prescription.java (hunk 5 above)
Comment (Bm2pires) [reply src/main/java/Models/Prescription/Prescription.java:13]: should be changed to a date type now in the class

File: src/main/java/Models/Prescription/Prescription.java (hunk 6 above)
Comment (Bm2pires) [reply src/main/java/Models/Prescription/Prescription.java:14]: changed to reflect date type being in the prescription object.

File: src/main/java/Models/Prescription/PrescriptionDAO.java (hunk 8 above)
Comment (Bm2pires) [reply src/main/java/Models/Prescription/PrescriptionDAO.java:28]: changed to date types

File: src/main/java/Controllers/IssuePrescriptionServlet.java (hunk 3 above)
Comment (Bm2pires) [reply src/main/java/Controllers/IssuePrescriptionServlet.java:41]: Done!

File: src/main/java/Utils/Tables.java (hunk 9 above)
Comment (Bm2pires) [reply src/main/java/Utils/Tables.java]: Commented out!

File: src/main/java/Models/Prescription/Prescription.java (hunk 7 above)
Comment (Bm2pires) [reply src/main/java/Models/Prescription/Prescription.java]: Resolved!

**Review by CharlievWilliams — APPROVED** [review]
Looks good!

**Review by Tom-Goring — APPROVED** [review]
